File: controllers/bg_remove_controller.py
from fastapi import APIRouter, UploadFile, File, Form
from fastapi.responses import JSONResponse
from services.bg_remove_service import bg_remove_service

router = APIRouter(tags=["Image Background Remove"])

@router.post("/bgremove")
async def bg_remove(
    image: UploadFile = File(...),
    bg_type: str = Form(...),  # transparent | color | image
    bg_color: str | None = Form(None),
    bg_image: UploadFile | None = File(None),
):
    try:
        result = await bg_remove_service(
            image=image,
            bg_type=bg_type,
            bg_color=bg_color,
            bg_image=bg_image,
        )

        return {
            "status": "success",
            "preview": result["preview"],
            "final": result["final"],
        }

    except Exception as e:
        print("❌ BG REMOVE ERROR:", e)
        return JSONResponse(400, {"message": str(e)})

File: src/modules/bg-remove/bgRemove.controller.ts
import type { Request, Response } from "express";
import { bgRemoveSchema } from "./bgRemove.validator.js";
import { processImage } from "./bgRemove.service.js";


export const bgRemove = async (req: Request, res: Response) => {
  try {
    const files = req.files as {
      [fieldname: string]: Express.Multer.File[];
    };

    const image = files?.image?.[0];
    const bgImage = files?.bg_image?.[0];

    if (!image) {
      return res.status(400).json({ message: "Image is required" });
    }

    if (!image.mimetype.startsWith("image/")) {
      return res.status(400).json({
        message: "Only image file is allowed"
      });
    }

    const parsed = bgRemoveSchema.safeParse(req.body);

    if (!parsed.success) {
      return res.status(400).json({
        message: parsed.error.issues[0].message
      });
    }

    if (parsed.data.bg_type === "image" && !bgImage) {
      return res.status(400).json({
        message: "Background image is required"
      });
    }

    const result = await processImage({
      image,
      bgType: parsed.data.bg_type,
      bgColor: parsed.data.bg_color,
      bgImage
    });

    return res.json(result);
  } catch (e) {
    console.error(e);
    return res.status(500).json({
      message: "Background remove failed"
    });
  }
};
